File: torchsl/csrc/multiview/mvpls.cpp
#include <vector>
#include <tuple>
#include <torch/script.h>
#include "multiview_helpers.h"

namespace torchsl {
std::tuple<torch::Tensor, torch::Tensor> mvpls(
    const std::vector<torch::Tensor>& Xs
    ) {
    const auto options = Xs[0].options().requires_grad(false);
    const int num_views = Xs.size();
    const int num_components = Xs[0].size(0);
    const int num_samples = num_views * num_components;
    const auto dims = torchsl::dimensions(Xs);

    auto I = torch::eye(num_components, options);
    auto B = torch::ones({ num_components, num_components }, options).div_(num_samples);

    auto SI = torch::eye(dims.sum().item<int>(), options);
    auto Sb = torchsl::multiview_covariance_matrix(dims, [&Xs, &I, &B] (int j, int r) {
        return Xs[j].t().mm(I - B).mm(Xs[r]);
    }, options);

    return std::make_tuple(SI, Sb);
}
}

File: torchsl/ops/multiview_subspace_learning/mvpls.py
import torch

import torchsl
from torchsl._extensions import _has_ops
from ._helpers import *

__all__ = ['mvpls']


# ===========================================================
# Multiview Partial Least Squared Regression
# ===========================================================
def mvpls(Xs):
    if _has_ops():
        return torchsl.ops.mvpls(Xs)

    options = dict(dtype=Xs[0].dtype, device=Xs[0].device)
    num_views = len(Xs)
    num_components = Xs[0].size(0)
    num_samples = num_views * num_components
    dims = dimensions(Xs)

    I = torch.eye(num_components, **options)
    B = torch.ones(num_components, num_components, **options) / num_samples

    SI = torch.eye(dims.sum(), **options)
    Sb = multiview_covariance_matrix(dims,
                                     lambda j, r: Xs[j].t().mm(I - B).mm(Xs[r]) if j != r
                                     else 0,
                                     options)

    return SI, Sb
